src/serialize.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import utils = require('./utils');
+import kernel = require('./kernel');
+
+import IKernelMsg = kernel.IKernelMsg;
+
+
+/**
+ * Deserialize and return the unpacked message.
+ */
+export
+function deserialize(data: ArrayBuffer | string): IKernelMsg {
+  var value: IKernelMsg;
+  if (typeof data === "string") {
+    value = JSON.parse(data);
+  } else {
+    value = deserializeBinary(data);
+  }
+  return value;
+}
+
+
+/**
+ * Serialize a kernel message for transport.
+ */
+export
+function serialize(msg: IKernelMsg): string | ArrayBuffer {
+  var value: string | ArrayBuffer;
+  if (msg.buffers && msg.buffers.length) {
+    value = serializeBinary(msg);
+  } else {
+    value = JSON.stringify(msg);
+  }
+  return value;
+}
+
+
+/**
+ * Deserialize a binary message to a Kernel Message.
+ */
+function deserializeBinary(buf: ArrayBuffer): IKernelMsg {
+  var data = new DataView(buf);
+  // read the header: 1 + nbufs 32b integers
+  var nbufs = data.getUint32(0);
+  var offsets: number[] = [];
+  for (var i = 1; i <= nbufs; i++) {
+    offsets.push(data.getUint32(i * 4));
+  }
+  var json_bytes = new Uint8Array(buf.slice(offsets[0], offsets[1]));
+  var msg = JSON.parse((new TextDecoder('utf8')).decode(json_bytes));
+  // the remaining chunks are stored as DataViews in msg.buffers
+  msg.buffers = [];
+  for (var i = 1; i < nbufs; i++) {
+    var start = offsets[i];
+    var stop = offsets[i + 1] || buf.byteLength;
+    msg.buffers.push(new DataView(buf.slice(start, stop)));
+  }
+  return msg;
+}
+
+
+/**
+ * Implement the binary serialization protocol.
+ * Serialize Kernel message to ArrayBuffer.
+ */
+function serializeBinary(msg: IKernelMsg): ArrayBuffer {
+  var offsets: number[] = [];
+  var buffers: ArrayBuffer[] = [];
+  var encoder = new TextEncoder('utf8');
+  var json_utf8 = encoder.encode(JSON.stringify(msg, replacer));
+  buffers.push(Array.prototype.slice.call(json_utf8));
+  for (var i = 0; i < msg.buffers.length; i++) {
+    // msg.buffers elements could be either views or ArrayBuffers
+    // buffers elements are ArrayBuffers
+    var b: any = msg.buffers[i];
+    buffers.push(b instanceof ArrayBuffer ? b : b.buffer);
+  }
+  var nbufs = buffers.length;
+  offsets.push(4 * (nbufs + 1));
+  for (i = 0; i + 1 < buffers.length; i++) {
+    offsets.push(offsets[offsets.length - 1] + buffers[i].byteLength);
```

Comment:
But this line is definitely wrong. When `i = 0` `buffers[i]` is an `Array`, and arrays don't have a `byteLength` property.